Source/bmalloc/bmalloc/VMAllocate.h

```diff
@@ -229,42 +292,94 @@ inline void vmAllocatePhysicalPages(void* p, size_t vmSize)
 #endif
 #endif
 }
+#else
+// Windows
+inline size_t vmPageSize()
+{
+    static size_t cached;
+    if (!cached) {
+        SYSTEM_INFO systemInfo;
+        GetSystemInfo(&systemInfo);
+        cached = systemInfo.dwPageSize;
+    }
+    return cached;
+}
 
-// Returns how much memory you would commit/decommit had you called
-// vmDeallocate/AllocatePhysicalPagesSloppy with p and size.
-inline size_t physicalPageSizeSloppy(void* p, size_t size)
+inline size_t vmPageSizePhysical()
 {
-    char* begin = roundUpToMultipleOf(vmPageSizePhysical(), static_cast<char*>(p));
-    char* end = roundDownToMultipleOf(vmPageSizePhysical(), static_cast<char*>(p) + size);
+    static size_t cached;
+    if (!cached) {
+        SYSTEM_INFO systemInfo;
+        GetSystemInfo(&systemInfo);
+        cached = systemInfo.dwPageSize;
+    }
+    return cached;
+}
 
-    if (begin >= end)
-        return 0;
-    return end - begin;
+static inline DWORD protection(bool writable, bool executable)
+{
+    return executable ?
+        (writable ? PAGE_EXECUTE_READWRITE : PAGE_EXECUTE_READ) :
+        (writable ? PAGE_READWRITE : PAGE_READONLY);
 }
 
-// Trims requests that are un-page-aligned.
-inline void vmDeallocatePhysicalPagesSloppy(void* p, size_t size)
+inline void* tryVMAllocate(size_t vmSize, VMTag usage)
 {
-    char* begin = roundUpToMultipleOf(vmPageSizePhysical(), static_cast<char*>(p));
-    char* end = roundDownToMultipleOf(vmPageSizePhysical(), static_cast<char*>(p) + size);
+    BUNUSED_PARAM(usage);
+    const bool writable = true;
+    const bool executable = true;
+    return VirtualAlloc(nullptr, vmSize, MEM_RESERVE, protection(writable, executable));
+}
 
-    if (begin >= end)
+inline void vmDeallocate(void* p, size_t vmSize)
+{
+    // https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc
+    // Use MEM_RESET to purge physical pages at timing of OS's preference. This is aligned to
+    // madvise MADV_FREE / MADV_FREE_REUSABLE.
+    // https://devblogs.microsoft.com/oldnewthing/20170113-00/?p=95185
+    // > The fact that MEM_RESET does not remove the page from the working set is not actually mentioned
+    // > in the documentation for the MEM_RESET flag. Instead, it’s mentioned in the documentation for
+    // > the Offer­Virtual­Memory function, and in a sort of backhanded way
+    // So, we need VirtualUnlock call.
```

Comment:
I'm not a Windows expert but this behavior as documented is incorrect. vmDeallocate needs to free the virtual address space, not just the physical pages.

Reply:
Yeah looks like you're right, I lifted this from OSAllocatorWin without paying close enough attention to the desired behavior of vmDeallocate.